docs/proposals/deployment.md

```diff
@@ -0,0 +1,172 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/deployment.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment
+
+## Abstract
+
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin of
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string][string]
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The percentage of pod replicas that should be updated in each interval.
+  // If nil, a fixed value of 1 pod is used.
+  UpdatePercent *int
```

Comment:
I'm trying to craft intent-oriented policies as opposed to mechanism-oriented knobs, like max in flight, rate limit, etc.

For instance, rather than just a fixed interval or rate limit, the amount of time a pod needs to be ready before considering it "available" seems superior.

Yes, maxUnavailabilityPercentage would bound the number of pods updated simultaneously, since gone == unavailable. 

I'd make the percentage apply to the max of the old and new sizes, but not include the surge amount.

I don't think of the percentage as a batch size, but as a cap. Progress should be incremental.

Maybe we don't need the policy.

Examples:

maxUnavailabilityPercentage=30 and maxSurgePercentage=0: The size would be reduced by 30% immediately. New pods would be created as the old ones disappeared. Once the new pods were ready for long enough to be considered available, the old RC could be downsized further, followed by increasing the new RC more.

maxUnavailabilityPercentage=0 and maxSurgePercentage = 30: The size of the new RC would be set to 30% of the original/desired immediately. The old RC would be downsized only after the new pods were ready for long enough to be considered available.

maxUnavailabilityPercentage=10 and maxSurgePercentage = 20: The size of the new RC would be set to 30% of the original/desired immediately. The old RC would be downsized by 10% immediately. Progress thereafter would proceed as in the cases above.